src/D2L.Bmx/UpdateHandler.cs

```diff
@@ -0,0 +1,156 @@
+using System.Formats.Tar;
+using System.IO.Compression;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace D2L.Bmx;
+
+internal class UpdateHandler {
+
+	public async Task HandleAsync() {
+		if( !Directory.Exists( BmxPaths.OLD_BMX_VERSIONS_PATH ) ) {
+			try {
+				Directory.CreateDirectory( BmxPaths.OLD_BMX_VERSIONS_PATH );
+			} catch( Exception ex ) {
+				throw new BmxException( "Failed to initialize temporary BMX file directory (~/.bmx/temp)", ex );
+			}
+		}
+
+		using var httpClient = new HttpClient();
+		GithubRelease? releaseData = await GithubRelease.GetLatestReleaseDataAsync();
+		Version? latestVersion = releaseData?.GetReleaseVersion();
+		if( latestVersion is null ) {
+			throw new BmxException( "Failed to find the latest version of BMX." );
+		}
+
+		var localVersion = Assembly.GetExecutingAssembly().GetName().Version;
+		if( latestVersion <= localVersion ) {
+			Console.WriteLine( $"You already have the latest version {latestVersion}" );
+			return;
+		}
+
+		string archiveName = GetOSFileName();
+		string? downloadUrl = releaseData?.Assets?.FirstOrDefault( a => a.Name == archiveName )?.BrowserDownloadUrl;
+		if( string.IsNullOrWhiteSpace( downloadUrl ) ) {
+			throw new BmxException( "Failed to find the download URL of the latest BMX" );
+		}
+
+		string? currentFilePath = Environment.ProcessPath;
+		if( string.IsNullOrEmpty( currentFilePath ) ) {
+			throw new BmxException( "BMX could not update" );
+		}
+
+		string downloadPath = Path.GetTempFileName();
+		try {
+			var archiveRes = await httpClient.GetAsync( downloadUrl, HttpCompletionOption.ResponseHeadersRead );
+			using var fs = new FileStream( downloadPath, FileMode.Create, FileAccess.Write, FileShare.None );
+			await archiveRes.Content.CopyToAsync( fs );
+			await fs.FlushAsync();
+		} catch( Exception ex ) {
+			throw new BmxException( "Failed to download the update", ex );
+		}
+
+		string extractFolder = Path.Combine( Path.GetTempPath(), Path.GetRandomFileName() );
+		try {
+			Directory.CreateDirectory( extractFolder );
+		} catch( Exception ex ) {
+			File.Delete( downloadPath );
+			throw new BmxException( "Failed to initialize temporary folder for downloaded file", ex );
+		}
+
+		string currentDirectory = Path.GetDirectoryName( currentFilePath )!;
+		long backupPathTimeStamp = DateTime.Now.Millisecond;
+		string backupPath = Path.Join( BmxPaths.OLD_BMX_VERSIONS_PATH, $"bmx-v{localVersion}-{backupPathTimeStamp}-old.bak" );
+		try {
+			string extension = Path.GetExtension( downloadUrl );
+
+			if( extension.Equals( ".zip", StringComparison.OrdinalIgnoreCase ) ) {
+				ExtractZipFile( downloadPath, extractFolder );
+			} else if( extension.Equals( ".gz", StringComparison.OrdinalIgnoreCase ) ) {
+				ExtractTarGzipFile( downloadPath, extractFolder );
+			} else {
+				throw new Exception( "Unknown archive type" );
+			}
+		} catch( Exception ex ) {
+			Directory.Delete( extractFolder, true );
```

Comment:
Boolean literals should almost always be named explicitly
```suggestion
			Directory.Delete( extractFolder, recursive: true );
```
similarly for other `Directory.Delete` calls